core/pouch/migrations.js

```diff
@@ -0,0 +1,244 @@
+  baseSchemaVersion: SchemaVersion,
+  targetSchemaVersion: SchemaVersion,
+  description: string,
+  affectedDocs: (Metadata[]) => Metadata[],
+  run: (Metadata[]) => Metadata[]
+}
+*/
+
+const SCHEMA_DOC_ID = '_local/schema'
+const SCHEMA_INITIAL_VERSION = 0
+const INITIAL_SCHEMA = {
+  _id: SCHEMA_DOC_ID,
+  version: SCHEMA_INITIAL_VERSION
+}
+const MIGRATION_RESULT_NOOP = 'MigrationNoop'
+const MIGRATION_RESULT_COMPLETE = 'MigrationComplete'
+const MIGRATION_RESULT_FAILED = 'MigrationFailed'
+
+const migrations /*: Migration[] */ = []
+
+function migrationNoop() {
+  return { type: MIGRATION_RESULT_NOOP, errors: [] }
+}
+
+async function currentSchemaVersion(
+  db /*: PouchDB */
+) /*: Promise<SchemaVersion> */ {
+  const schema = (await safeGet(db, SCHEMA_DOC_ID)) || INITIAL_SCHEMA
+  return schema.version || SCHEMA_INITIAL_VERSION
+}
+
+async function updateSchemaVersion(
+  version /*: SchemaVersion */,
+  db /*: PouchDB */
+) {
+  const schema = (await safeGet(db, SCHEMA_DOC_ID)) || INITIAL_SCHEMA
+  return await db.put({ ...schema, version })
+}
+
+async function migrate(
+  migration /*: Migration */,
+  pouch /*: Pouch */
+) /*: Promise<MigrationResult> */ {
+  if ((await currentSchemaVersion(pouch.db)) !== migration.baseSchemaVersion) {
+    return migrationNoop()
+  } else {
+    const migrationDB = createDB(await migrationDBPath(pouch.db))
+
+    let result /*: MigrationResult */
+    try {
+      const docs = await pouch.byRecursivePathAsync('')
+      const affectedDocs = migration.affectedDocs(docs)
+      const migratedDocs = migration.run(affectedDocs)
+
+      if (migratedDocs.length) {
+        await replicateDB(pouch.db, migrationDB)
+        result = await save(migratedDocs, migrationDB)
+      } else {
+        result = migrationNoop()
+      }
+
+      switch (result.type) {
+        case MIGRATION_RESULT_NOOP:
+          await updateSchemaVersion(migration.targetSchemaVersion, pouch.db)
+          break
+        case MIGRATION_RESULT_COMPLETE:
+          await updateSchemaVersion(migration.targetSchemaVersion, migrationDB)
+          await pouch.resetDatabaseAsync()
+          await replicateDB(migrationDB, pouch.db)
+          break
+      }
+    } catch (err) {
+      result = { type: MIGRATION_RESULT_FAILED, errors: [err.toString()] }
+    } finally {
+      migrationDB.destroy()
+    }
+
+    return result
+  }
+}
+
+function createDB(name /*: string */) /*: PouchDB */ {
+  return new PouchDB(name).on('error', err => {
+    throw err
+  })
+}
+
+async function migrationDBPath(
+  originalDB /*: PouchDB */
+) /*: Promise<string> */ {
+  const date = new Date()
+  const dateString = `${date.getFullYear()}-${date.getMonth() +
+    1}-${date.getDate()}`
+  const originalDBInfo = await originalDB.info()
+  return `${originalDBInfo.db_name}-migration-${dateString}`
+}
+
+async function replicateDB(fromDB /*: PouchDB */, toDB /*: PouchDB */) {
+  return new Promise(async (resolve, reject) => {
+    fromDB.replicate
```

Comment:
`replicate` doesn't delete the documents in `toDB` that are not in `fromDB`. I think it may cause some issues for migrations that change the `_id` of documents. For example, we may want to play a migration to add a `/` at the start of the `_id` of each doc to fix the issue for files starting with an underscore.

Reply:
There shouldn't be any issues here.
When we create the migration DB, `toDB` is empty.
When we replace the user's DB with the content of the migration DB, we have previously reset the user's DB so `toDB` is empty in this situation as well.

Reply:
It's probably my mistake, but I don't see where the user's DB is reset

Reply:
It's done in the `migrate` method: https://github.com/cozy-labs/cozy-desktop/pull/1682/files/c5b3d72078270f2201bdf986c5a3da278005c30d#diff-71833bc7fc5def55987ed47e8e557c16R98